### PR TITLE
WT-6692 Handle scenario where imported timestamps are newer than current DB's timestamps

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
I think this change does what we want.
The main decisions here that aren't currently part of the technical design are:

- We're comparing against `oldest`. Using `stable` is another possibility but since we're not importing history store data, timestamped reads could return unexpected results.
- We're checking the metadata string instead of inspecting the root page directly.

### CODE REVIEW

File: src/schema/schema_create.c [hunk 1]
```diff
@@ -44,6 +44,49 @@ __wt_direct_io_size_check(
...
+              "%s: import found an aggregated durable timestamp newer than the current stable "
+              "timestamp, newest_start_durable_ts=%" PRIu64 ", stable_ts=%" PRIu64,
+              uri, ckpt->ta.newest_start_durable_ts, txn_global->stable_timestamp);
+
+        if (ckpt->ta.oldest_start_ts > txn_global->stable_timestamp)
+            WT_ERR_MSG(session, EINVAL,
+              "%s: import found an aggregated commit timestamp newer than the current stable "
+              "timestamp, oldest_start_ts=%" PRIu64 ", stable_ts=%" PRIu64,
+              uri, ckpt->ta.oldest_start_ts, txn_global->stable_timestamp);
+    }
```
Comment (tetsuo-cpp) [inline src/schema/schema_create.c:82]: I think we also need to add checks for the stop timestamps like this:
```
if (ckpt->ta.newest_stop_ts != WT_TS_MAX && ckpt->ta.newest_stop_ts > txn_global->stable-timestamp)
```

Comment (tetsuo-cpp) [reply src/schema/schema_create.c:96]: 5e08f5c

Comment (tetsuo-cpp) [reply src/schema/schema_create.c:96]: I think there's something wrong here. I think that in the case of stop timestamps, the aggregated time isn't enough information.

In the case that not every record has a tombstone, `newest_stop_ts` will be `WT_TS_MAX`. In that case, it's not possible whether there's any non-MAX timestamp greater than stable.

Comment (kommiharibabu) [reply src/schema/schema_create.c:96]: @tetsuo-cpp I saw the above comment, I am not sure of the full details. But there is an aggregated `newest_durable_stop_ts` that stores the newest tombstone that is present on the table. Maybe it is useful for you?

Comment (tetsuo-cpp) [reply src/schema/schema_create.c:96]: @kommiharibabu 
Maybe. If there is a record with no tombstone, will `newest_durable_stop_ts` be `WT_TS_MAX`? Or is it the newest non-max timestamp.

Comment (kommiharibabu) [reply src/schema/schema_create.c:96]: The `newest_durable_stop_ts` will always be a non-max timestamp even if the record with no tombstone.

Comment (tetsuo-cpp) [reply src/schema/schema_create.c:96]: Had a chat with @kommiharibabu.
Looks like the behaviour of `newest_durable_stop_ts` tells us what we need.

File: src/schema/schema_create.c [hunk 2]
```diff
@@ -136,12 +189,20 @@ __create_file(WT_SESSION_IMPL *session, const char *uri, bool exclusive, const c
...
+    } else {
         /* Create the file. */
         WT_ERR(__wt_block_manager_create(session, filename, allocsize));
 
-    if (WT_META_TRACKING(session))
-        WT_ERR(__wt_meta_track_fileop(session, NULL, uri));
+        /*
+         * Track the creation of this file.
+         *
+         * If something down the line fails, we're going to need to roll this back. Specifically do
```
Comment (tetsuo-cpp) [inline src/schema/schema_create.c:199]: This part is not specific to this change but it's something I noticed when running my new test.

If the import failed after this point, we were previously wiping the data file off the disk. I've made this change to only do this when we're not importing.

File: test/suite/test_import01.py [hunk 3]
```diff
@@ -64,9 +72,9 @@ def check_record(self, uri, key, value, ts):
     def check(self, uri, keys, values, ts):
```
Comment (tetsuo-cpp) [inline test/suite/test_import01.py:72]: Since the timestamps have to be behind oldest in order to import, it doesn't make sense to read from a range of timestamps so I use the same timestamp for all reads.

We could just do away with the timestamp usage in `check` and just read without one.

File: src/schema/schema_create.c [hunk 4]
```diff
@@ -44,6 +44,58 @@ __wt_direct_io_size_check(
     return (0);
 }
 
+/*
+ * __check_imported_ts --
+ *     Check the aggregated timestamps for each checkpoint in a file that we've imported. We're not
+ *     allowed to import files with timestamps ahead of our oldest timestamp since a subsequent
```
[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply src/schema/schema_create.c:50]: This comment is just saying that importing timestamps later than oldest shouldn't be allowed because:

### If an update is later than oldest, then it's within the queryable window of read timestamps and some of the results we return may be inconsistent with the old database.

In the old database with oldest=30, stable=50:
U6 (ts=40) -> U5 (ts=30)

In the new database with oldest=30, stable=50:
U6 (ts=40)

In this case, a read at timestamp 35 will return different results before and after import since we're not bringing across history. We don't have this issue if the updates must be behind oldest since we can't query behind it in the new database.

### If an update is later than stable, then rollback to stable after import will be destructive.

In the old database with oldest=30, stable=50:
U6 (ts=40) -> U5 (ts=30)

In the new database with oldest=30, stable=30:
U6 (ts=40)

A rollback to stable will remove this record which may not be expected.

In summary, we need to ensure that updates are earlier than stable in order to avoid the second issue. And we need to ensure that updates are earlier than oldest to avoid the first issue.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/schema/schema_create.c:50]: Understood. Thanks for the explanation. we are not importing more than oldest timestamp, do we need to mention RTS here without mentioning a stable timestamp? Also, do we need to check whether the stable timestamp is less than the oldest timestamp?

[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply src/schema/schema_create.c:50]: > Understood. Thanks for the explanation. we are not importing more than oldest timestamp, do we need to mention RTS here without mentioning a stable timestamp? Also, do we need to check whether the stable timestamp is less than the oldest timestamp?

When can that situation occur?
That doesn't make sense to me. I've always thought of stable as always being greater than the oldest.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/schema/schema_create.c:50]: I think it is possible to set only the oldest timestamp but not stable, I feel WT still allows it. Maybe it is not a proper usage, so no need to worry.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by agorrod** [comment]
@tetsuo-cpp is this waiting on me?

**Comment by tetsuo-cpp** [comment]
> @tetsuo-cpp is this waiting on me?

Yes, it's waiting on both you and @tammybailey.

File: test/suite/test_import01.py (hunk 3 above)
Comment (tammybailey) [reply test/suite/test_import01.py:72]: We should remove the timestamp variable from the check functions if you are sure there's no reason for it.

Comment (tetsuo-cpp) [reply test/suite/test_import01.py]: 4960bf6

**Review by tammybailey — APPROVED** [review]
lgtm.

**Comment by tetsuo-cpp** [comment]
@agorrod 
Friendly ping.
Could you please take a look soon? This PR makes some test changes that affect my other branches.

**Review by agorrod — APPROVED** [review]
Sorry for sitting on this for so long @tetsuo-cpp lgtm.
